Closes #2735 derive extreme event examples

manciniedoardo opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF and a PNG version of it to the same folder. (The PNG version can be created by taking a screenshot of the PDF version.)
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

manciniedoardo commented:
@rossfarrugia @bundfussr I've given this my best shot 🥲 - please review and let me know if anything is too verbose or out of place. I've found it hard to think of simple examples to this function that do not trigger most of the arguments!

Also (sorry @bundfussr I know we just talked about it but please could you help with the `join_type` explanation in CBOR? <y brain is fried and I didn't want to write something that is just wrong...!

bundfussr commented:
> @rossfarrugia @bundfussr I've given this my best shot 🥲 - please review and let me know if anything is too verbose or out of place. I've found it hard to think of simple examples to this function that do not trigger most of the arguments!
> 
> Also (sorry @bundfussr I know we just talked about it but please could you help with the `join_type` explanation in CBOR? <y brain is fried and I didn't want to write something that is just wrong...!

I wonder if we should add some simple examples to `derive_vars_joined_summary()` to demonstrate how `join_type`, `first_cond_lower`, and `first_cond_upper` work. Something like
```
> derive_vars_joined_summary(
+   mydata,
+   dataset_add = mydata,
+   order = exprs(DAY),
+   join_type = "after",
+   new_vars = exprs(USED_DAYS = paste(DAY, collapse = ","))
+ )
# A tibble: 5 × 2
    DAY USED_DAYS
  <dbl> <chr>    
1     1 2,3,4,5  
2     2 3,4,5    
3     3 4,5      
4     4 5        
5     5 NA       
> derive_vars_joined_summary(
+   mydata,
+   dataset_add = mydata,
+   order = exprs(DAY),
+   join_type = "after",
+   first_cond_upper = DAY.join == 4,
+   new_vars = exprs(USED_DAYS = paste(DAY, collapse = ","))
+ )
# A tibble: 5 × 2
    DAY USED_DAYS
  <dbl> <chr>    
1     1 2,3,4    
2     2 3,4      
3     3 4        
4     4 NA       
5     5 NA  
```
@rossfarrugia , @manciniedoardo , what do you think?

manciniedoardo commented:
> > @rossfarrugia @bundfussr I've given this my best shot 🥲 - please review and let me know if anything is too verbose or out of place. I've found it hard to think of simple examples to this function that do not trigger most of the arguments!
> > Also (sorry @bundfussr I know we just talked about it but please could you help with the `join_type` explanation in CBOR? <y brain is fried and I didn't want to write something that is just wrong...!
> 
> I wonder if we should add some simple examples to `derive_vars_joined_summary()` to demonstrate how `join_type`, `first_cond_lower`, and `first_cond_upper` work. Something like
> 
> ```
> > derive_vars_joined_summary(
> +   mydata,
> +   dataset_add = mydata,
> +   order = exprs(DAY),
> +   join_type = "after",
> +   new_vars = exprs(USED_DAYS = paste(DAY, collapse = ","))
> + )
> # A tibble: 5 × 2
>     DAY USED_DAYS
>   <dbl> <chr>    
> 1     1 2,3,4,5  
> 2     2 3,4,5    
> 3     3 4,5      
> 4     4 5        
> 5     5 NA       
> > derive_vars_joined_summary(
> +   mydata,
> +   dataset_add = mydata,
> +   order = exprs(DAY),
> +   join_type = "after",
> +   first_cond_upper = DAY.join == 4,
> +   new_vars = exprs(USED_DAYS = paste(DAY, collapse = ","))
> + )
> # A tibble: 5 × 2
>     DAY USED_DAYS
>   <dbl> <chr>    
> 1     1 2,3,4    
> 2     2 3,4      
> 3     3 4        
> 4     4 NA       
> 5     5 NA  
> ```
> 
> @rossfarrugia , @manciniedoardo , what do you think?

I think that would be nice

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#'   knowing whether a subject's Alkaline Phosphatase has ever not
+#'   been below twice the upper limit of normal range. If so, i.e. if
+#'   `CRIT1FL` is `NA`, we are interested in the record for the first
+#'    time this occurs, and if not, we wish to retain the last record.
```
Could we add something extra here to explain why this is a different case, as when I first went through this i kept thinking why is this example not first as it seems the simplest? but then i caught the need to use `mode = "last"` within the event.

I'm thinking a small extra sentence like: 
```
So in this case now we need to vary our usage of the `mode` argument dependent on the `event()`.
```

manciniedoardo replied:
agree, will add

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
 #' )
 #'
-#' # Use different mode by event
+#' @caption Specifying different arguments across `event()` object
```
```suggestion
#' @caption Specifying different arguments across `event()` objects
```

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#'    time this occurs, and if not, we wish to retain the last record.
+#'
+#'  - In first `event()`, since we simply seek the first time that
+#'    `CRIT1FL` is `NA`, it's enough to specify the `condition`.
```
Only because of the `order` and `mode` that have been set in the `derive_extreme_event()` call itself which are taking the first occurrence by `AVISITN` here - should we clarify this? As it still needs to be applied this way by the user.

manciniedoardo replied:
you're right, i'd intended to explicitly say this but forgot. thanks

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#'  - In the second `event()`, we select the last record among the
+#'    full set of records where `CRIT1FL = "Y"` by additionally
+#'     specifying`mode = "last"` within the `event()`.
+#'
```
I think we should add a bullet to call out usage of `keep_source_vars = exprs(AVISITN)` here instead of the usual `exprs(everything)`, which makes sense here as you wouldn't want `CRIT1FL` populated for the new records.

manciniedoardo replied:
great point, will add.

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -331,6 +476,7 @@
 #' ) %>%
 #'   filter(PARAMCD == "CBOR")
 #'
+
```
Do we need a blank line here and above?

Also we don't have a example for `check_type` around handling duplicates here - i would consider linking to stefan's `derive_extreme_records()` example here? i did similar with `derive_vars_joined()` linking to `derive_vars_merged()`. These uniqueness checks do catch out users sometimes

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -199,9 +319,17 @@
+#' @caption A more complex example: Confirmed Best Overall Response
+#' @info The final example showcases a use of `derive_extreme_event()`
+#'   to calculate the Confirmed Best Overall Response (CBOR) in an
+#'   ADRS dataset. This example builds on all the previous ones and
```
```suggestion
#'   `ADRS` dataset. This example builds on all the previous ones and
```
Maybe also mention something like "as commonly used in oncology clinical trials"?

rossfarrugia replied:
@manciniedoardo there's a couple more cases of this later in the example bullets that also need to be updated

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -199,9 +319,17 @@
+#'   thus assumes a baseline level of confidence with
+#'   `derive_extreme_event()`.
+#'
+#'   The following ADSL and ADRS datasets will be used
```
```suggestion
#'   The following `ADSL` and `ADRS` datasets will be used
```

rossfarrugia replied:
General nice convention i feel - i see same in other places so will stop commenting this and feel free to update the other cases if you agree

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -252,6 +380,37 @@
+#'
+#'   - For the Confirmed Response (CR), for each `"CR"` record in the original ADRS
+#'     dataset that will be identified by the first part of the `condition` argument
+#'     (`AVALC == "CR"`) we need to use the `first_cond_upper` argument to limit the
```
```suggestion
#'     (`AVALC == "CR"`), we need to use the `first_cond_upper` argument to limit the
```

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -252,6 +380,37 @@
+#'     In particular, within `condition` we use `all()` to check that all observations
+#'     are either `"CR"` or `"NE"`, and `count_vals()` to ensure at most one is
+#'     `"NE"`. Note that here we have selected `join_type = "after"` because
+#'     <insert explanation here, explaining why join_type = "all" changes patient 1 to a "PR">.
```
Definitely best @bundfussr provides this text

rossfarrugia replied:
i'll approve as from my side i'm all good with the PR, but of course this bit needs to be updated before merging

manciniedoardo replied:
I think i've got my head around this now - i've added an explanation if you'd like to check @bundfussr @rossfarrugia

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -252,6 +380,37 @@
+#'
+#'  - The Partial Response (PR), is very similar; with the difference being that the
+#'    first portion of `condition` now references `"PR"` and `first_cond_upper`
+#'    accepts a confirmatory PR or CR 28 days later. Note that now we must add
```
```suggestion
#'    accepts a confirmatory `"PR"` or `"CR"` 28 days later. Note that now we must add
```

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -252,6 +380,37 @@
+#'    first portion of `condition` now references `"PR"` and `first_cond_upper`
+#'    accepts a confirmatory PR or CR 28 days later. Note that now we must add
+#'    `"PR"` as an option within the `all()` condition to account for confirmatory
+#'    PRs.
```
```suggestion
#'    "`PR`"s.
```

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -252,6 +380,37 @@
+#'    events are simpler and just require `event()` calls.
+#'
+#'  - Finally, we use a catch-all `event()` with `condition = TRUE` and
+#'    `dataset_name = adsl` to identify those subjects who do not appear in ADRS
```
Explain you also need `source_datasets = list(adsl = adsl)` in the function call as you're using a different input dataset to `dataset` for this `event()`.

manciniedoardo replied:
agree, will add.

manciniedoardo commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
 #' @export
 #'
-#' @examples
+#' @examplesx
```
you're right, done, thanks

rossfarrugia replied:
@manciniedoardo you missed this for some arguments: dataset, by_vars, events, tmp_event_nr_var, source_datasets, set_values_to. following Stefan's example i've been adding it to all arguments

manciniedoardo replied:
sorry, looks like i missed the inherited params - added now 🤦

rossfarrugia replied:
thanks! it's still missing for events, tmp_event_nr_var and source_datasets too which are not inherited, but then i looked at mine and Stefan's earlier PRs and noticed we missed adding it for some arguments too. i don't remember the rationale but maybe it's that we're fine to leave them out when it's very obvious from the description 🤷‍♂️ could be worth someone sanity checking all these that we've added for consistency before release, as we don't really want common arguments documented inconsistently across functions and it might just be simplest rule to have a Permitted entry for every argument.

manciniedoardo replied:
Yeah i think we need to discuss as a team, but might be easiest to put permitteds everywhere. for now i will add to this PR

manciniedoardo commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
 #'
-#' @examples
+#' @examplesx
+#'
```
will solve while updating

manciniedoardo commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
 #' derive_extreme_event(
-#'   adqs,
+#'   adqs1,
 #'   by_vars = exprs(USUBJID),
```
It's a required argument insofar that you have to always feed an expression, but it can be empty - see this (slightly silly) example where I left `by_vars = exprs()`  and `check_type = "none"` to avoid a warning for duplicates.
```
library(tibble)
library(dplyr)
library(lubridate)

adqs1 <- tribble(
  ~USUBJID, ~PARAMCD,         ~AVALC,        ~ADY,
  "1",      "NO SLEEP",       "N",              1,
  "1",      "WAKE UP 3X",     "N",              2,
  "2",      "NO SLEEP",       "N",              1,
  "2",      "WAKE UP 3X",     "Y",              2,
  "2",      "WAKE UP 3X",     "Y",              3,
  "3",      "NO SLEEP",       NA_character_,    1
) %>%
mutate(STUDYID = "AB42")

derive_extreme_event(
  adqs1,
  by_vars = exprs(),
  events = list(
    event(
      condition = PARAMCD == "NO SLEEP" & AVALC == "Y",
      set_values_to = exprs(AVALC = "No sleep", AVAL = 1)
    ),
    event(
      condition = PARAMCD == "WAKE UP 3X" & AVALC == "Y",
      set_values_to = exprs(AVALC = "Waking up three times", AVAL = 2)
    ),
    event(
      condition = all(AVALC == "N"),
      set_values_to = exprs(
        AVALC = "No sleeping problems", AVAL = 3
      )
    ),
    event(
      condition = TRUE,
      set_values_to = exprs(AVALC = "Missing", AVAL = 99)
    )
  ),
  tmp_event_nr_var = event_nr,
  order = exprs(event_nr, desc(ADY)),
  mode = "first",
  set_values_to = exprs(
    PARAMCD = "WSP",
    PARAM = "Worst Sleeping Problem"
  ),
  keep_source_vars = exprs(everything()),
  check_type = "none"
) %>%
select(-STUDYID)
```
Returns:
```
# A tibble: 7 × 6
  USUBJID PARAMCD    AVALC                   ADY  AVAL PARAM                 
  <chr>   <chr>      <chr>                 <dbl> <dbl> <chr>                 
1 1       NO SLEEP   N                         1    NA NA                    
2 1       WAKE UP 3X N                         2    NA NA                    
3 2       NO SLEEP   N                         1    NA NA                    
4 2       WAKE UP 3X Y                         2    NA NA                    
5 2       WAKE UP 3X Y                         3    NA NA                    
6 3       NO SLEEP   NA                        1    NA NA                    
7 2       WSP        Waking up three times     3     2 Worst Sleeping Problem
```
Personally I don't see the need to showcase something like this because in an ADaM sense it would be very rare to look across patients.

rossfarrugia replied:
ah makes sense, glad it works when empty `by_vars` are provided just in the rare case ever needed. agree though it doesn't warrant another example. i was trying to think how to break down the simplest possible first example, but as you say it wouldn't be a likely realistic usage so lets leave it as is

manciniedoardo commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#'    select `join_type = "all"`.
+#'  - Since we are now passing an `event_joined` object to
+#'    `derive_extreme_event()`, we also need to specify the `source_datasets`
+#'    object so as to indicate which dataset we wish to use for the join.
```
yes, thanks

manciniedoardo commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#' @caption Specifying different arguments across `event()` object
+#' @info Here we consider a Hy's law use case. We are interested in
+#'   knowing whether a subject's Alkaline Phosphatase has ever not
+#'   been below twice the upper limit of normal range. If so, i.e. if
```
agree, will change

manciniedoardo commented:
@rossfarrugia I still need to figure out why the roxygen2 check is failing but think I have addressed everything else. Thanks for the review!

rossfarrugia commented:
> @rossfarrugia I still need to figure out why the roxygen2 check is failing but think I have addressed everything else. Thanks for the review!

No idea either. i tried from my side re-running `roxygen2::roxygenize('.')` and `devtools::document()` and there seems no changes needed. might be a CI bug, but lets see if @bundfussr spots anything when he reviews this one

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#'
+#' - The groups for which new records are added are specified by the `by_vars`
+#'   argument. Here for each *subject* a record should be added. Thus
+#'   `by_vars = exprs(USUBJID)` is specified.
```
@manciniedoardo the CBOR example at the end needs to get the following added:

```
 %>%
select(-STUDYID)
```

rossfarrugia replied:
actually i'd also drop `ADTC` as it doesn't add anything when `ADT` is already there and its just more for the user to read

manciniedoardo replied:
added, thanks

bundfussr commented:
> > @rossfarrugia I still need to figure out why the roxygen2 check is failing but think I have addressed everything else. Thanks for the review!
> 
> No idea either. i tried from my side re-running `roxygen2::roxygenize('.')` and `devtools::document()` and there seems no changes needed. might be a CI bug, but lets see if @bundfussr spots anything when he reviews this one

I think the issue is that the man checks starts from scratch while in your home you usually generate the `.md` files which were changed. The conflict warning occurs only at the first `library()` call where a conflict occurs. I.e., it depends on which `.md` files are generated and the order of them.

For now, I would add `warn.conflicts = FALSE` to the library calls. Once we have implemented https://github.com/pharmaverse/admiraldev/issues/489, we could suppress the start up messages.

rossfarrugia commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +106,181 @@
+#'    and finally, because we wish to compare all records with each other, we
+#'    select `join_type = "all"`.
+#'  - Since we are now passing an `event_joined` object to
+#'    `derive_extreme_event()`, we also need to specify the `source_datasets`
```
ok thanks, the explanation works

rossfarrugia commented on man/roxygen/rdx_meta.R:
```diff
@@ -14,7 +16,8 @@ list(
+    var_list = "list of variables created by `exprs()`, e.g., `exprs(USUBJID, VISIT)`",
     var = "a variable",
-    var_list = "list of variables created by `exprs()`, e.g., `exprs(USUBJID, VISIT)`"
+    event = "an `event()` or `event_joined()` object"
```
minor comment but easier to maintain if this is alphabetised

manciniedoardo replied:
done

bundfussr commented on R/derive_extreme_event.R:
```diff
@@ -197,11 +361,20 @@
-#' # Derive confirmed best overall response (using event_joined())
-#' # CR - complete response, PR - partial response, SD - stable disease
-#' # NE - not evaluable, PD - progressive disease
+#' @caption A more complex example: Confirmed Best Overall Response
```
Do you want to show how to derive CBOR or how to add events requiring confirmation or how `cond_first_upper`, `join_type = "after"` and `dataset_name`/`source_datasets` work?
Depending on this we may need to adjust the title.

manciniedoardo replied:
I've added the arguments in the title, as i think it's mainly the last two options

bundfussr reviewed (changes requested):
For me the examples are clear and well written. However, I'm already familiar with the function and the examples. Thus I'm not sure if it the same for readers who have never used the function. Maybe they would prefer simpler examples?

manciniedoardo commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +130,200 @@
+#'    and finally, because we wish to compare all records with each other, we
+#'    select `join_type = "all"`.
+#'  - Since we are now passing an `event_joined` object to
+#'    `derive_extreme_event()`, it is good practice to explicitly specify the
+#'    `source_datasets` argument so as to indicate which dataset we wish to
+#'    use for the join. Note that in this case, it is not strictly needed
+#'    as even in the `event_joined()` object we are using the same source
+#'    dataset as for the general call to `derive_extreme_event()`, i.e.
+#'    `adqs2`, but this is not always the case (see the final CBOR example).
```
you're right. I've just removed this bit, as we discuss this in the final section of the CBOR example anyway.

manciniedoardo commented:
> For me the examples are clear and well written. However, I'm already familiar with the function and the examples. Thus I'm not sure if it the same for readers who have never used the function. Maybe they would prefer simpler examples?

@bundfussr thanks for the review - think i've addressed your comments. Personally I think the simpler examples are likely to be further from potential applications of the function to real studies, so i think it's ok to expect the user to have to do a little bit of studying to understand how this works - it's a complex function after all.

bundfussr commented on R/derive_extreme_event.R:
```diff
@@ -156,37 +286,62 @@
 #' ) %>%
 #'   mutate(
 #'     PARAMCD = "ALKPH",
-#'     PARAM = "Alkaline Phosphatase (U/L)"
```
Should be
```
#'     PARAMCD = "ALT",
#'     PARAM = "ALT (U/L)"
```
to be consistent with `CRIT1`.

manciniedoardo replied:
updated thanks

bundfussr commented on R/derive_extreme_event.R:
```diff
@@ -195,13 +350,23 @@
 #'   keep_source_vars = exprs(AVISITN),
 #'   set_values_to = exprs(
 #'     PARAMCD = "ALK2",
```
```suggestion
#'     PARAMCD = "ALT2",
```

manciniedoardo replied:
updated thanks

bundfussr commented on R/derive_extreme_event.R:
```diff
@@ -156,37 +286,62 @@
-#'   "2",             2, NA_character_,
-#'   "2",             3, "Y",
-#'   "2",             4, NA_character_
+#'   ~USUBJID, ~AVISITN,              ~CRIT1,      ~CRIT1FL,
+#'   "1",             1, "ALT > 2 times ULN", NA_character_,
+#'   "1",             2, "ALT > 2 times ULN", NA_character_,
+#'   "2",             1, "ALT > 2 times ULN", NA_character_,
+#'   "2",             2, "ALT > 2 times ULN", "Y",
+#'   "2",             3, "ALT > 2 times ULN", NA_character_,
+#'   "2",             4, "ALT > 2 times ULN", "Y"
```
This is not ADaM compliant. Either `CRIT1` should be `NA` if `CRIT1FL` is `NA` or `CRIT1FL` should be `"N"` if the criterion is not fulfilled.

manciniedoardo replied:
good point updated thanks

bundfussr commented on R/derive_extreme_event.R:
```diff
@@ -106,61 +130,200 @@
+#'
+#' - The groups for which new records are added are specified by the `by_vars`
+#'   argument. Here for each *subject* a record should be added. Thus
+#'   `by_vars = exprs(USUBJID)` is specified.
```
@manciniedoardo , if you want to keep `STUDYID`, you should update `by_vars = exprs(USUBJID)` to `by_vars = exprs(STUDYID, USUBJID)` in the text to be consistent with the code.

manciniedoardo commented:
![image](https://github.com/user-attachments/assets/bdb36296-7363-4eb5-9b0d-ec6335a42754)

thanks @bundfussr @rossfarrugia for the reviews!
